### PR TITLE
Support Add to Existing App use case

### PR DESCRIPTION
**Pull request by junying1**
The current implementation assumes the plugin is used in a pure Flutter app. It casts the app's rootViewController to FlutterViewController.

This supports using this plugin in an "add to app" use case. The implementation does not assume the rootViewController is an instance of FlutterViewController, but rather starts with the presented view controller, and navigates up the stack to the root controller. It also handles in the case of a non-modal screen, in which case it would be a UINavigationController. We could probably just use the presented view controller, rather than navigating up the stack to the root to use as the session context provider.

### CODE REVIEW

File: ios/Classes/SwiftFlutterWebAuthPlugin.swift [hunk 1]
```diff
@@ -46,8 +46,18 @@ public class SwiftFlutterWebAuthPlugin: NSObject, FlutterPlugin {
                 let session = ASWebAuthenticationSession(url: url, callbackURLScheme: callbackURLScheme, completionHandler: completionHandler)
 
                 if #available(iOS 13, *) {
-                    guard let provider = UIApplication.shared.delegate?.window??.rootViewController as? FlutterViewController else {
-                        result(FlutterError(code: "FAILED", message: "Failed to aquire root FlutterViewController" , details: nil))
+                    if var topController = UIApplication.shared.keyWindow?.rootViewController {
```
Comment (LinusU) [inline ios/Classes/SwiftFlutterWebAuthPlugin.swift:49]: There is still code that references the `provider` variable on line 55/65 so I don't think that these changes will compile 😅

File: ios/Classes/SwiftFlutterWebAuthPlugin.swift [hunk 2]
```diff
@@ -46,8 +46,18 @@ public class SwiftFlutterWebAuthPlugin: NSObject, FlutterPlugin {
...
+                        while let presentedViewController = topController.presentedViewController {
+                            topController = presentedViewController
+                        }
+                        if topController is UINavigationController {
+                            var viewControllerProvider = (topController as! UINavigationController).visibleViewController!
+                            session.presentationContextProvider = viewControllerProvider as! ASWebAuthenticationPresentationContextProviding
+                        } else {
+                            session.presentationContextProvider = topController as! ASWebAuthenticationPresentationContextProviding
+                        }
```
Comment (LinusU) [inline ios/Classes/SwiftFlutterWebAuthPlugin.swift:58]: I'm a bit weary about the force unwrapping here. Could we change it out for checked casts instead?

**Comment by LinusU** [comment]
Sorry for the slow review on this one 🙈 

Would you be able to tell me how to test this out? Would be perfect if you have an example project that demonstrates the problem ☺️

File: ios/Classes/SwiftFlutterWebAuthPlugin.swift (hunk 2 above)
Comment (junying1) [reply ios/Classes/SwiftFlutterWebAuthPlugin.swift]: Ok changed and pushed the commit.

File: ios/Classes/SwiftFlutterWebAuthPlugin.swift (hunk 1 above)
Comment (junying1) [reply ios/Classes/SwiftFlutterWebAuthPlugin.swift:49]: Yeah missed it when I merged master.

**Comment by junying1** [comment]
> Sorry for the slow review on this one 🙈
> 
> Would you be able to tell me how to test this out? Would be perfect if you have an example project that demonstrates the problem ☺️

I don't have a readily available public project for you to test. You need to create a Flutter module that uses the plugin, then create a natives iOS project to embed the Flutter module. If you have a project you use to test this plugin you can turn it into a module and embed in a Xcode project. The instruction is here: https://flutter.dev/docs/development/add-to-app/ios/project-setup. I have my version running in 2 commercial projects.

File: ios/Classes/SwiftFlutterWebAuthPlugin.swift [hunk 3]
```diff
@@ -71,6 +89,10 @@ public class SwiftFlutterWebAuthPlugin: NSObject, FlutterPlugin {
             result(FlutterMethodNotImplemented)
         }
     }
+
+    private func returnError(result: @escaping FlutterResult) {
```
Comment (LinusU) [inline ios/Classes/SwiftFlutterWebAuthPlugin.swift:93]: The function name is very generic "return error", but the actual error is very specific. I think that the best approach would be to make a `fileprivate extension` to `FlutterError` that adds a `aquireRootViewControllerFailed` getter that returns the error:

```swift
fileprivate extension FlutterError {
    var aquireRootViewControllerFailed: FlutterError {
        return FlutterError(code: "AQUIRE_ROOT_VIEW_CONTROLLER_FAILED", message: "Failed to aquire root view controller" , details: nil)
    }
```

Then you can remove this function and just do:

```diff
- returnError(result: result)
+ result(FlutterError.aquireRootViewControllerFailed)
```

File: ios/Classes/SwiftFlutterWebAuthPlugin.swift [hunk 4]
```diff
@@ -46,13 +46,31 @@ public class SwiftFlutterWebAuthPlugin: NSObject, FlutterPlugin {
                 let session = ASWebAuthenticationSession(url: url, callbackURLScheme: callbackURLScheme, completionHandler: completionHandler)
 
                 if #available(iOS 13, *) {
-                    guard let provider = UIApplication.shared.delegate?.window??.rootViewController as? FlutterViewController else {
-                        result(FlutterError(code: "FAILED", message: "Failed to aquire root FlutterViewController" , details: nil))
+                    if var topController = UIApplication.shared.keyWindow?.rootViewController {
```
Comment (LinusU) [inline ios/Classes/SwiftFlutterWebAuthPlugin.swift:49]: Since this `if` statement has a returning else condition, could you change it to a `guard`?

**Comment by LinusU** [comment]
Just some small changes, otherwise looks great 👍 

I understand if you don't have the motivation to work on this since I've been so slow to respond, sorry about that! I'll try to apply the suggested changes myself and merge this if you don't feel like it ☺️

**Comment by junying1** [comment]
> Just some small changes, otherwise looks great 👍
> 
> I understand if you don't have the motivation to work on this since I've been so slow to respond, sorry about that! I'll try to apply the suggested changes myself and merge this if you don't feel like it ☺️

Thanks for reviewing. I agree with the changes. I actually had to change my computer so don't have this repo readily available. If you get to it before I do please go ahead and change and merge! Appreciate it.

**Comment by LinusU** [comment]
A version of this was merged in https://github.com/LinusU/flutter_web_auth/pull/106 🎉

**Pull request closed without merging**